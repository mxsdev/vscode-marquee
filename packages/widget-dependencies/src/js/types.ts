import type vscode from 'vscode'
import type { DependencyType } from 'src/types'

export type JsPackageManager = 'npm'|'yarn'|'pnpm'

export interface JsPackageDependency {
  name: string
  dependencyType?: DependencyType
  /**
   * Query version in package.json
   */
  packageVersion: string
  /**
   * Version in node_modules
   */
  actualVersion?: string
  homepage?: string
  repository?: string
}

export interface LocalJsPackage extends JsPackageMeta {
  isWorkspace?: boolean
  name: string
  version: string
  dependencies: Record<string, JsPackageDependency>
  uri: vscode.Uri
  workspaces?: string[]
  manager: JsPackageManager
}

export type LocalJsProjects = Record<string, LocalJsPackage>

export type RemoteRegistry = Record<string, JsPackageMeta>

export interface JsPackageMeta {
  latestVersion?: string
  homepage?: string
  repository?: string
}

export interface NPMOutdatedInfo {
  packageId: string
  current?: string
  wanted?: string
  latest?: string
  workspace?: string
  packageType?: 'devDependencies'|'dependencies'
  url?: string
}

export type NPMOutdatedHead = (
  | 'Package'
  | 'Current'
  | 'Wanted'
  | 'Latest'
  | 'Workspace'
  | 'Package Type'
  | 'URL'
)[]

export type NPMOutdatedColumn = NPMOutdatedHead[number]

export interface NPMOutdatedData {
  body: string[][]
  head: NPMOutdatedHead
}