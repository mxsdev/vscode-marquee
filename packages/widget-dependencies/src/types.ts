import type { ContextProperties } from '@vscode-marquee/utils'
import type vscode from 'vscode'
export type { RawPackument } from 'query-registry'

export interface State {
  dependencies: DisplayedDependency[]
  loading: boolean
  capabilities: DependencyProviderCapabilities
  event: (EventsObj & { id: EventId }) | {}
}

export interface Configuration {
  /* General */
  showUpToDate: boolean
  autoRefresh: boolean

  /* JS */
  jsRegistry: string
  prefersPnpm: boolean
}

export interface Events {
  refreshDependencies: {}
  updateDependency: { packageId: string, toVersion: string, workspace?: string }
  updateAllDependencies: {}
  removeDependency: { packageId: string, workspace?: string, isRootWorkspace: boolean }
}

export interface Context extends ContextProperties<Configuration & State> {
  _refreshDependencies(): void
  _updateDependency(dep: DisplayedDependency, toVersion: string): void
  _updateAllDependencies(): void
  _removeDependency(dep: DisplayedDependency): void
}

export interface DependencyProvider extends vscode.Disposable {
  loadDependencies(): Promise<DisplayedDependency[]|undefined>

  upgradeDependency?: (
    packageId: string,
    upgradeTo: string,
    workspace: string|undefined
  ) => Promise<void>

  upgradeAllDependencies?: () => Promise<void>

  deleteDependency?: (
    packageId: string,
    workspace: string|undefined,
    isRootWorkspace: boolean
  ) => Promise<void>

  /**
   * Whether or not this provider explicitly sets if this dependency needs
   * an upgrade via the `needsUpgrade` prop of `DisplayedDependency`
   *
   * If this is false, the client determines whether a package needs an update
   * by comparing the current and latest semantic versions
   */
  explicitNeedsUpgrade?: () => Promise<boolean>
}

export interface TerminalProvider extends vscode.Disposable {
  getOrCreateTerminal(cwd: vscode.Uri): vscode.Terminal
}

export type DependencyType = 'normal'|'dev'

export interface DisplayedDependency {
  name: string

  versions: {
    current?: string
    wanted?: string
    latest?: string
    query?: string
  },

  project?: string

  isRootWorkspace: boolean

  dependencyType?: DependencyType

  url?: string

  needsUpgrade?: boolean
}

export interface DependencyProviderCapabilities {
  deleteDependency: boolean
  upgradeDependency: boolean
  upgradeAllDependencies: boolean
  explicitNeedsUpgrade: boolean
}

export type EventsObj<K extends keyof Events = keyof Events>
  = K extends string ? { type: K, payload: Events[K] } : never

export type EventId = number
