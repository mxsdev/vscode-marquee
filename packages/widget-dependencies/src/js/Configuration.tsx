import { FormControlLabel, Checkbox, TextField } from '@mui/material'
import React, { useContext } from 'react'
import DependencyContext from '../Context'

const Configuration = () => {
  const {
    prefersPnpm,
    setPrefersPnpm,
    jsRegistry,
    setJsRegistry
  } = useContext(DependencyContext)
  
  return (
    <>
      <FormControlLabel 
        control={
          <Checkbox 
            checked={prefersPnpm}
            onChange={(e) => setPrefersPnpm(e.target.checked)}
          />
        }
        label="Prefer pnpm"
      />

      <TextField 
        id="widget-dependencies-config-js-registry" 
        label="Registry" 
        variant="filled"

        value={jsRegistry}
        onChange={(e) => setJsRegistry(e.target.value)}

        placeholder="Default registry..."
      />
    </>
  )
}

export default Configuration