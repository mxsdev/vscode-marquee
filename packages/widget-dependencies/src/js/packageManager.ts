import type { PackageJSON } from 'query-registry'
import vscode  from 'vscode'
import type { 
  JsPackageManager, 
  LocalJsPackage, 
  LocalJsProjects, 
  NPMOutdatedColumn, 
  NPMOutdatedData, 
  NPMOutdatedInfo 
} from './types'

declare const IS_WEB_BUNDLE: boolean

export async function tryGetPackageJson (
  uri: vscode.Uri,
): Promise<PackageJSON|undefined> {
  const pkgJsonPath = vscode.Uri.joinPath(uri, 'package.json')
    
  let file: string

  try {
    const rawBytes = (await vscode.workspace.fs.readFile(pkgJsonPath))
    // buffer.toString() not supported on web
    file = IS_WEB_BUNDLE ? new TextDecoder().decode(rawBytes) : rawBytes.toString()
  } catch(e) {
    // no package.json found
    return undefined
  }

  let pkgJson: PackageJSON

  try {
    pkgJson = JSON.parse(file)
  } catch(e) {
    throw new Error('Unable to serialize package.json')
  }

  if (typeof pkgJson !== 'object' 
    || typeof pkgJson.name !== 'string'
    || typeof pkgJson.version !== 'string'
  ) { 
    throw new Error('Invalid package.json')
  }

  return pkgJson
}

export async function tryGetJsProject (
  uri: vscode.Uri,
  parent?: LocalJsPackage,
  prefersPnpm?: boolean,
): Promise<LocalJsPackage|undefined> {
  const pkgJson = await tryGetPackageJson(uri)
  if(!pkgJson) { return undefined }

  const repositoryUrl = typeof pkgJson.repository === 'string' && pkgJson.repository
  const workspaces = pkgJson.workspaces as string[]|{ projects: string[] }|undefined

  const pkg: LocalJsPackage = {
    name: pkgJson.name,
    version: pkgJson.version,
    dependencies: {},
    isWorkspace: !!parent,
    workspaces: Array.isArray(workspaces) ? workspaces : workspaces?.projects,
    uri,
    manager: parent?.manager ?? await getPackageManager(uri, prefersPnpm),
    homepage: pkgJson.homepage,
    repository: repositoryUrl || undefined
  }

  const dependenciesZipped = [
    ...Object.entries(pkgJson.dependencies ?? {}).map(pkg => ([pkg, 'normal'] as const)),
    ...Object.entries(pkgJson.devDependencies ?? {}).map(pkg => ([pkg, 'dev'] as const))
  ]

  await Promise.all(
    dependenciesZipped.map(async ([[name, version], dependencyType]) => {
      const localJson = await tryGetPackageJson(
        vscode.Uri.joinPath(pkg.uri, 'node_modules', ...name.split('/'))
      ) ?? (parent && await tryGetPackageJson(
        vscode.Uri.joinPath(parent.uri, 'node_modules', ...name.split('/'))
      ))

      const repo = typeof localJson?.repository === 'string' ? localJson?.repository : undefined

      pkg.dependencies[name] = {
        name,
        packageVersion: version,
        actualVersion: localJson?.version,
        dependencyType,
        homepage: localJson?.homepage,
        repository: repo
      }
    })
  )

  return pkg

  async function getPackageManager (
    uri: vscode.Uri,
    prefersPnpm?: boolean,
  ): Promise<JsPackageManager> {
    try {
      await vscode.workspace.fs.stat(vscode.Uri.joinPath(uri, 'yarn.lock'))
      return 'yarn'
    } catch(e) { }

    return prefersPnpm ? 'pnpm' : 'npm'
  }
}

export async function findWorkspaces (
  rootProject: LocalJsPackage,
): Promise<LocalJsProjects> {
  const projects: LocalJsProjects = {
    [rootProject.name]: rootProject,
  }

  // apply glob patterns in workspaces
  const candidateUris = (await Promise.all(
    rootProject.workspaces?.map(workspacePath => 
      vscode.workspace.findFiles(
        new vscode.RelativePattern(rootProject.uri, workspacePath)
      )
    ) ?? []
  )).flatMap(x => x)
  
  // filter for only the folders, and add to projects
  await Promise.all(
    candidateUris.map(async (relativeUri) => {
      const uri = vscode.Uri.joinPath(rootProject.uri, relativeUri.fsPath)

      const { type } = await vscode.workspace.fs.stat(uri)

      if(type !== vscode.FileType.Directory) { return }

      const pkg = await tryGetJsProject(uri, rootProject)
      if(!pkg) { return }

      projects[pkg.name] = pkg
    })
  )

  return projects
}

export async function packageManagerCmdOutdated (
  pkg: LocalJsPackage,
): Promise<NPMOutdatedInfo[]|undefined> {
  const output = await packageManagerCmd(
    pkg, 
    'outdated',
    '--json',
  )

  if(!output) { return undefined }

  const lines = output.split('\n')

  const table = lines
    .find((line) => {
      try {
        const json = JSON.parse(line)
        return json.type === 'table'
      } catch(e) {
        return false
      }
    })
  
  if(!table) { return undefined }

  let data: NPMOutdatedData

  try {
    data = JSON.parse(table).data
  } catch(e) {
    return undefined
  }

  // invert header
  const idxMap = data.head.reduce(
    (prev, col, i) => {
      prev[col] = i
      return prev
    }, 
    {} as Record<NPMOutdatedColumn, number>
  )

  // Need packageId
  if(!('Package' in idxMap)) {
    return undefined
  }

  const fromRow = (from: string[], col: NPMOutdatedColumn) => {
    if (col in idxMap) { 
      return from[idxMap[col]] 
    }

    return undefined
  }

  return data.body.map(row => ({
    packageId: fromRow(row, 'Package')!,
    current: fromRow(row, 'Current'),
    latest: fromRow(row, 'Latest'),
    packageType: fromRow(row, 'Package Type') as NPMOutdatedInfo['packageType'],
    url: fromRow(row, 'URL'),
    wanted: fromRow(row, 'Wanted'),
    workspace: fromRow(row, 'Workspace')
  }))
}

export async function packageManagerCmd (
  pkg: LocalJsPackage,
  ...args: string[]
): Promise<string|undefined> {
  if(IS_WEB_BUNDLE) { return undefined }

  const { exec } = require('child_process') as typeof import('node:child_process')
  
  return await new Promise<string|undefined>((res, rej) => {
    const proc = exec(
      [pkg.manager, ...args].join(' '),
      { cwd: pkg.uri.fsPath },
      (err,stdout,stderr) => {
        if(stdout) {
          res(stdout)
        }

        if(err) {
          rej(err)
        }

        if(stderr) {
          rej(stderr)
        }
      }
    )

    proc.on('close', () => res(undefined))
  })
}