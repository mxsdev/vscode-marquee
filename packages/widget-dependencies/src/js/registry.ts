import { getFetch, urlJoin } from '../util'
import { RawPackument } from '../types'
import { DEFAULT_JS_REGISTRY } from '../constants'
import { LocalJsProjects, RemoteRegistry } from './types'
import { Logger } from '@vscode-marquee/utils/build/logger'

/**
 * Go through project/workspaces and pull package metadata for each dependency
 * from npm/remote registry provided by `registryUrl`
 */
export async function pullRemoteRegistry (
  projects: LocalJsProjects,
  registryUrl?: string
): Promise<RemoteRegistry> {
  const registry: RemoteRegistry = {}
  
  await Promise.all(
    Object.values(projects)
      .flatMap(repo => Object.values(repo.dependencies).map(async dep => {
        if(!projects[dep.name]) {
          const packument = await getRawPackument(dep.name, registryUrl)
          if(!packument) { return } // package does not exist
          
          const { versions, homepage, repository } = packument

          const version = packument['dist-tags'].latest

          const latest = versions[version]
          if(!latest) { return } // package has no releases

          const repositoryUrl = typeof repository === 'string' && repository

          registry[dep.name] = { 
            latestVersion: latest.version,
            homepage,
            repository: repositoryUrl || undefined
          }
        } else {
          registry[dep.name] = projects[dep.name]
        }
      }))
  )

  return registry
}

export async function getRawPackument (
  packageId: string, 
  registry: string|undefined
): Promise<RawPackument|undefined> {
  registry ||= DEFAULT_JS_REGISTRY
  const url = urlJoin(registry, packageId)

  const fetch = getFetch()

  return await fetch(url)
    .then(res => res.json() as Promise<RawPackument>)
    .catch((e) => { 
      Logger.error('Failed to fetch package', e)
      return undefined
    })
}