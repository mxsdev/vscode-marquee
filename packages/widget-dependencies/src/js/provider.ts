import vscode from 'vscode'
import { Configuration, DependencyProvider, DisplayedDependency, TerminalProvider } from 'src/types'
import { pullRemoteRegistry } from './registry'
import { findWorkspaces, packageManagerCmdOutdated, tryGetJsProject } from './packageManager'
import { JsPackageManager, LocalJsPackage } from './types'

declare const IS_WEB_BUNDLE: boolean

export class JsDependencyProvider implements DependencyProvider {
  private packageManager: JsPackageManager = 'npm'
  private disposables: vscode.Disposable[] = []
  
  constructor (
    private readonly config: Configuration,
    private terminalProvider: TerminalProvider,
    refreshDependencies: () => void,
  ) { 
    const watcher = vscode.workspace.createFileSystemWatcher(
      '**/package.json'
    )
    
    this.disposables.push(
      watcher
    )

    this.disposables.push(
      watcher.onDidChange(() => {
        refreshDependencies()
      })
    )
  }

  async loadDependencies (): Promise<DisplayedDependency[]|undefined> {
    const pkg = await this.getPackageInfo()
    if(!pkg) { return undefined }
  
    const projects = await findWorkspaces(pkg)

    if (!IS_WEB_BUNDLE) {
      const data = await packageManagerCmdOutdated(pkg) ?? []

      const upgradeableDepMemo: Record<string, Set<string>|undefined> = {}
      
      const upgradeableDeps: DisplayedDependency[] = await Promise.all(data.map(
        async ({
          packageId,
          current,
          wanted,
          latest,
          workspace,
          packageType,
          url
        }) => {
          const workspaceId = workspace ?? pkg.name
          
          if(!upgradeableDepMemo[workspaceId]) {
            upgradeableDepMemo[workspaceId] = new Set()
          }
          
          upgradeableDepMemo[workspaceId]?.add(packageId)

          const hasWorkspaces = (pkg.workspaces?.length ?? 0) > 0

          return ({
            name: packageId,
            versions: {
              current,
              wanted,
              latest,
              query: pkg.dependencies[packageId]?.packageVersion
            },
            dependencyType: packageType === 'dependencies' ? 'normal'
              : packageType === 'devDependencies' ? 'dev'
                : undefined,
            project: workspace,
            url,
            needsUpgrade: true,
            isRootWorkspace: hasWorkspaces && (!workspace || workspace === pkg.name)
          })
        }
      ))

      const upToDateDeps = this.config.showUpToDate && Object.values(projects)
        .flatMap(repo => 
          Object.values(repo.dependencies)
            .filter(dep => !upgradeableDepMemo[repo.name]?.has(dep.name))
            .map(dep => {
              const hasWorkspaces = (repo.workspaces?.length ?? 0) > 0

              return ({
                name: dep.name,
                versions: {
                  current: dep.actualVersion,
                  query: dep.packageVersion,
                  latest: dep.actualVersion,
                },
                ...repo.isWorkspace && { project: repo.name },
                dependencyType: dep.dependencyType,
                url: dep.homepage ?? dep.repository,
                isRootWorkspace: hasWorkspaces
              })
            })
        )

      return [
        ...upgradeableDeps,
        ...upToDateDeps || []
      ]
    } else {
      const remoteRegistry = await pullRemoteRegistry(projects, this.config.jsRegistry)

      return Object.values(projects)
        .flatMap(
          repo => 
            Object.values(repo.dependencies)
              .filter(dep => dep.name in remoteRegistry)
              .map(dep => {
                const remoteDep = remoteRegistry[dep.name]!

                const hasWorkspaces = (repo.workspaces?.length ?? 0) > 0

                return ({
                  name: dep.name,
                  versions: {
                    current: dep.actualVersion,
                    query: dep.packageVersion,
                    latest: remoteDep.latestVersion ?? undefined,
                  },
                  ...repo.isWorkspace && { project: repo.name },
                  dependencyType: dep.dependencyType,
                  url: remoteDep.homepage ?? remoteDep.repository,
                  isRootWorkspace: hasWorkspaces
                })
              })
        )
    }
  }

  async upgradeDependency (
    packageId: string, 
    toVersion: string, 
    workspace: string|undefined
  ) {
    this.packageManagerCommand(
      ['upgrade', `${packageId}@${toVersion}`],
      workspace
    )
  }

  async upgradeAllDependencies () {
    this.packageManagerCommand(['upgrade'])
  }

  async deleteDependency (
    packageId: string, 
    workspace: string|undefined,
    isRootWorkspace: boolean = false
  ) {
    this.packageManagerCommand(
      ['remove', packageId],
      workspace,
      isRootWorkspace
    )
  }

  async explicitNeedsUpgrade () {
    return !IS_WEB_BUNDLE
  }

  async getPackageInfo (): Promise<LocalJsPackage|undefined> {
    const { workspaceFolders } = vscode.workspace
    
    if(!workspaceFolders || workspaceFolders.length === 0) { return undefined }
    
    const workspacePath = workspaceFolders[0].uri
    const pkg = await tryGetJsProject(workspacePath, undefined, this.config.prefersPnpm)

    if(!pkg) { return undefined }

    this.packageManager = pkg.manager

    return pkg
  }

  packageManagerCommand (
    args: string[],
    workspace?: string,
    rootWorkspaceFlag = false
  ) {
    args = [
      ...rootWorkspaceFlag ? ['-W'] : [],
      ...(workspace && !rootWorkspaceFlag) ? ['workspace', workspace] : [],
      ...args
    ]
    
    const terminal = this.terminalProvider.getOrCreateTerminal()

    terminal.sendText(
      [ this.packageManager, ...args ].join(' '),
      true
    )

    terminal.show(true)
  }

  dispose () { 
    this.disposables.forEach(d => d.dispose())
  }
}