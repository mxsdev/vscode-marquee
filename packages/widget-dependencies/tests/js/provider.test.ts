import { JsDependencyProvider } from '../../src/js/provider'
import type { Configuration, DisplayedDependency, TerminalProvider } from '../../src/types'
import { NPMOutdatedInfo, RemoteRegistry } from '../../src/js/types'
import * as registry from '../../src/js/registry'
import vscode from 'vscode'
import vscodeMock from '../vscodeFsMock'
import projects from './projects'
import path from 'path'
import * as packageManager from '../../src/js/packageManager'

const terminal = {
  dispose: () => {},
  sendText: jest.fn(),
  show: jest.fn()
}

const terminalProvider = {
  dispose: () => {},
  getOrCreateTerminal: () => terminal
}

const refreshDependencies = jest.fn()

const remoteRegistry: RemoteRegistry = {
  '@types/node': {
    latestVersion: '18.11.18'
  },
  '@types/react': {
    latestVersion: '18.0.26'
  },
  '@emotion/react': {
    latestVersion: '11.10.5'
  },
  '@mui/material': {
    latestVersion: '5.11.4'
  },
  'typescript': {
    latestVersion: '4.9.4'
  },
  '@emotion/styled': {
    latestVersion: '11.10.5'
  },
  '@types/mocha': {
    latestVersion: '10.0.1'
  }
}

const packageManagerCmdOutdated = jest.fn<NPMOutdatedInfo[], any>(() => [])

const {
  _fsWatcherChangeDispose,
  _fsWatcherDispose
} = (vscode as unknown as typeof vscodeMock).workspace

const disposables = [
  _fsWatcherChangeDispose,
  _fsWatcherDispose
]

;(registry as any).pullRemoteRegistry = () => remoteRegistry
;(packageManager as any).packageManagerCmdOutdated = packageManagerCmdOutdated

jest.mock('vscode', () => jest.requireActual('../vscodeFsMock.ts'))

type ProviderConfig = Pick<Configuration, 'showUpToDate'> & { _webMode?: boolean }

describe('loadDependencies', () => {
  test('root project', providerTest(
    vscode.Uri.file(path.join(__dirname, '../../../../')),
    async (provider) => {
      const deps = await provider.loadDependencies()

      expect(deps).toBeTruthy()
      expect(deps).not.toHaveLength(0)
    }
  ))

  test('yarn project', providerTest(
    'yarn',
    async (provider) => {
      packageManagerCmdOutdated.mockImplementationOnce(() => [
        {
          packageId: '@mui/material',
          current: '5.10.13',
          latest: '5.11.4',
          wanted: '5.11.4',
          packageType: 'devDependencies',
          url: 'https://mui.com/material-ui/getting-started/overview/'
        },
        {
          packageId: '@types/react',
          current: '18.0.25',
          latest: '18.0.26',
          wanted: '18.0.26',
          packageType: 'devDependencies',
          url: 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react'
        }
      ])
      const deps = await provider.loadDependencies()

      expect(deps).toBeTruthy()
      expect(deps).toHaveLength(5)

      normalizeDependencyList(deps!)
      expect(deps).toMatchSnapshot()
    }
  ))

  test('yarn project with workspaces', providerTest(
    'yarnWorkspaces',
    async (provider) => {
      packageManagerCmdOutdated.mockImplementationOnce(() => [
        {
          packageId: '@types/node',
          current: '16.18.9',
          latest: '18.11.18',
          packageType: 'devDependencies',
          url: 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node',
          wanted: '16.18.9',
          workspace: 'yarn-project-workspace'
        }
      ])

      const deps = await provider.loadDependencies()

      expect(deps).toBeTruthy()
      expect(deps).toHaveLength(4)

      normalizeDependencyList(deps!)
      expect(deps).toMatchSnapshot()
    }
  ))
})

describe('upgradeDependency', () => {
  test('for root package', providerTest(
    'yarn',
    async (provider, uri) => {
      // must call so that package manager is set
      await provider.getPackageInfo()

      await provider.upgradeDependency('package', 'latest', undefined)

      expect(terminal.sendText).toBeCalledTimes(1)
      expect(terminal.sendText).toBeCalledWith(getPkgCmd('yarn upgrade package@latest', uri), true)
      expect(terminal.show).toBeCalledTimes(1)
      expect(terminal.show).toBeCalledWith(true)
    }
  ))

  test('for workspace package', providerTest(
    'yarnWorkspaces',
    async (provider, uri) => {
      // must call so that package manager is set
      await provider.getPackageInfo()

      await provider.upgradeDependency('package', 'latest', 'subspace')

      expect(terminal.sendText).toBeCalledTimes(1)
      expect(terminal.sendText).toBeCalledWith(getPkgCmd('yarn workspace subspace upgrade package@latest', uri), true)
      expect(terminal.show).toBeCalledTimes(1)
      expect(terminal.show).toBeCalledWith(true)
    }
  ))
})

describe('deleteDependency', () => {
  test('for root package', providerTest(
    'yarn',
    async (provider, uri) => {
      // must call so that package manager is set
      await provider.getPackageInfo()

      await provider.deleteDependency('package', undefined)

      expect(terminal.sendText).toBeCalledTimes(1)
      expect(terminal.sendText).toBeCalledWith(getPkgCmd('yarn remove package', uri), true)
      expect(terminal.show).toBeCalledTimes(1)
      expect(terminal.show).toBeCalledWith(true)
    }
  ))

  test('for workspace package', providerTest(
    'yarnWorkspaces',
    async (provider, uri) => {
      // must call so that package manager is set
      await provider.getPackageInfo()

      await provider.deleteDependency('package', 'subspace')

      expect(terminal.sendText).toBeCalledTimes(1)
      expect(terminal.sendText).toBeCalledWith(getPkgCmd('yarn workspace subspace remove package', uri), true)
      expect(terminal.show).toBeCalledTimes(1)
      expect(terminal.show).toBeCalledWith(true)
    }
  ))

  test('for workspace package root', providerTest(
    'yarnWorkspaces',
    async (provider, uri) => {
      // must call so that package manager is set
      await provider.getPackageInfo()

      await provider.deleteDependency('package', 'root', true)

      expect(terminal.sendText).toBeCalledTimes(1)
      expect(terminal.sendText).toBeCalledWith(getPkgCmd('yarn -W remove package', uri), true)
      expect(terminal.show).toBeCalledTimes(1)
      expect(terminal.show).toBeCalledWith(true)
    }
  ))
})

test('upgradeAllDependencies', providerTest(
  'yarn',
    async (provider, uri) => {
    // must call so that package manager is set
    await provider.getPackageInfo()

    await provider.upgradeAllDependencies()

    expect(terminal.sendText).toBeCalledTimes(1)
    expect(terminal.sendText).toBeCalledWith(getPkgCmd('yarn upgrade', uri), true)
    expect(terminal.show).toBeCalledTimes(1)
    expect(terminal.show).toBeCalledWith(true)
  }
))

function getPkgCmd (cmd: string, uri: vscode.Uri): string {
  return `${cmd} --cwd ${uri.fsPath}`
}

function providerTest (
  project: vscode.Uri | keyof typeof projects,
  cb: (provider: JsDependencyProvider, uri: vscode.Uri) => Promise<void>,
  config?: ProviderConfig,
) {
  return async () => {
    const uri = typeof project === 'object' ? project : vscode.Uri.file(projects[project])

    const provider = createProvider(
      uri,
      config
    )

    await cb?.(provider, uri)

    provider.dispose()

    disposables.forEach(d => {
      expect(d).toBeCalledTimes(1)
      d.mockClear()
    })

    terminal.sendText.mockClear()
    terminal.show.mockClear()
  }

  function createProvider (
    uri: vscode.Uri,
    config: ProviderConfig = { showUpToDate: true }
  ) {
    (vscode.workspace as any)._setWorkspaceFolder(uri)

    return new JsDependencyProvider(
      config as Configuration,
      terminalProvider as unknown as TerminalProvider,
      refreshDependencies
    )
  }
}

// Sorts list of `DisplayedDependency`s to ensure consistent appearance across
// test-runners in snapshots
function normalizeDependencyList (
  deps: DisplayedDependency[]
) {
  deps.sort((a, b) => {
    {
      const comp = a.name.localeCompare(b.name)
      if(comp !== 0) { return comp }
    }

    {
      const comp = a.project?.localeCompare(b.project ?? a.project) ?? 0
      return comp
    }
  })
}
