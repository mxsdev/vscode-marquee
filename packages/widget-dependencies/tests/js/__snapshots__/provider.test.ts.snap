// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadDependencies yarn project 1`] = `
[
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@emotion/react",
    "url": "https://github.com/emotion-js/emotion/tree/main/packages/react",
    "versions": {
      "current": "11.10.5",
      "latest": "11.10.5",
      "query": "^11.10.5",
    },
  },
  {
    "dependencyType": "normal",
    "isRootWorkspace": false,
    "name": "@emotion/styled",
    "url": "https://github.com/emotion-js/emotion/tree/main/packages/styled",
    "versions": {
      "current": "11.10.5",
      "latest": "11.10.5",
      "query": "^11.10.5",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@mui/material",
    "needsUpgrade": true,
    "project": undefined,
    "url": "https://mui.com/material-ui/getting-started/overview/",
    "versions": {
      "current": "5.10.13",
      "latest": "5.11.4",
      "query": "^5.10.13",
      "wanted": "5.11.4",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@types/react",
    "needsUpgrade": true,
    "project": undefined,
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react",
    "versions": {
      "current": "18.0.25",
      "latest": "18.0.26",
      "query": "^18.0.25",
      "wanted": "18.0.26",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "typescript",
    "url": "https://www.typescriptlang.org/",
    "versions": {
      "current": "4.9.4",
      "latest": "4.9.4",
      "query": "4.9.4",
    },
  },
]
`;

exports[`loadDependencies yarn project with workspaces 1`] = `
[
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@types/mocha",
    "project": "yarn-project-workspace-package1",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/mocha",
    "versions": {
      "current": "10.0.1",
      "latest": "10.0.1",
      "query": "^10.0.1",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": true,
    "name": "@types/node",
    "needsUpgrade": true,
    "project": "yarn-project-workspace",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node",
    "versions": {
      "current": "16.18.9",
      "latest": "18.11.18",
      "query": "16.18.9",
      "wanted": "16.18.9",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@types/node",
    "project": "yarn-project-workspace-package3",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node",
    "versions": {
      "current": "18.11.18",
      "latest": "18.11.18",
      "query": "^18.11.18",
    },
  },
  {
    "dependencyType": "normal",
    "isRootWorkspace": true,
    "name": "@types/react",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react",
    "versions": {
      "current": "18.0.26",
      "latest": "18.0.26",
      "query": "^18.0.26",
    },
  },
]
`;

exports[`loadDependencies yarn project with workspaces 2`] = `
[
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@types/mocha",
    "project": "yarn-project-workspace-package1",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/mocha",
    "versions": {
      "current": "10.0.1",
      "latest": "10.0.1",
      "query": "^10.0.1",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": true,
    "name": "@types/node",
    "needsUpgrade": true,
    "project": "yarn-project-workspace",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node",
    "versions": {
      "current": "16.18.9",
      "latest": "18.11.18",
      "query": "16.18.9",
      "wanted": "16.18.9",
    },
  },
  {
    "dependencyType": "dev",
    "isRootWorkspace": false,
    "name": "@types/node",
    "project": "yarn-project-workspace-package3",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node",
    "versions": {
      "current": "18.11.18",
      "latest": "18.11.18",
      "query": "^18.11.18",
    },
  },
  {
    "dependencyType": "normal",
    "isRootWorkspace": true,
    "name": "@types/react",
    "url": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react",
    "versions": {
      "current": "18.0.26",
      "latest": "18.0.26",
      "query": "^18.0.26",
    },
  },
]
`;
